tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -4261,32 +4261,37 @@ static GlobalVariable *CreateStaticGlobal(llvm::Module *M, GlobalVariable *GV) {
 static bool CreateWriteEnabledStaticGlobals(llvm::Module *M) {
   std::vector<GlobalVariable *> worklist;
   for (GlobalVariable &GV : M->globals()) {
-    if (!GV.isConstant()) {
+    if (!GV.isConstant() && GV.getLinkage() != GlobalValue::LinkageTypes::InternalLinkage) {
       worklist.emplace_back(&GV);
       GV.setConstant(true);
     }
   }
 
+  BasicBlock *mainFunctionEntryBlock = nullptr;
   for (GlobalVariable *GV : worklist) {
-    std::set<BasicBlock *> entryBlocks;
-    for (User *U : GV->users()) {
-      if (Instruction *I = dyn_cast<Instruction>(U)) {
-        BasicBlock *entryBlock =
-            &(I->getParent()->getParent()->getEntryBlock());
-        if (entryBlocks.find(entryBlock) == entryBlocks.end())
-          entryBlocks.insert(entryBlock);
+    // find the entry block of main() function
+    if (!mainFunctionEntryBlock) {
+      for (User *U : GV->users()) {
+        if (Instruction *I = dyn_cast<Instruction>(U)) {
+          Function *F = I->getParent()->getParent();
+          if (F->getName() == "main") {
```

Comment:
Main entry function is not necessarily called "main".  You should pass the function in to CreateStaticGlobal() from FinishCodeGen.  You can get the main function by calling GetEntryFunction() on the HLModule.  And then you don't need to search for users that are in this function either, since there may be no users in main function, but you still want to insert memcpy there.